### PR TITLE
Issue 929

### PR DESCRIPTION
**Pull request by prb112**

### CODE REVIEW

File: fhir-persistence-schema/src/main/resources/add_any_resource.sql [hunk 1]
```diff
@@ -1,5 +1,5 @@
 -------------------------------------------------------------------------------
--- (C) Copyright IBM Corp. 2016, 2019
+-- (C) Copyright IBM Corp. 2016, 2020
 --
 -- SPDX-License-Identifier: Apache-2.0
```
Comment (albertwang-ibm) [inline fhir-persistence-schema/src/main/resources/add_any_resource.sql:4]: only change copyright year?

File: fhir-persistence-schema/src/main/resources/add_parameter_name.sql [hunk 2]
```diff
@@ -1,5 +1,5 @@
 -------------------------------------------------------------------------------
--- (C) Copyright IBM Corp. 2016
+-- (C) Copyright IBM Corp. 2016, 2020
 --
 -- SPDX-License-Identifier: Apache-2.0
```
Comment (albertwang-ibm) [inline fhir-persistence-schema/src/main/resources/add_parameter_name.sql:4]: only update copyright year?

File: fhir-persistence-schema/src/main/resources/add_resource_template.sql [hunk 3]
```diff
@@ -58,11 +58,6 @@ BEGIN
   SELECT t.tenant_id INTO {{SCHEMA_NAME}}.session_tenant
     FROM {{SCHEMA_NAME}}.tenants t
    WHERE t.tenant_name = p_tenant_data_id;
-  
-  -- Stop right here if we don't have a valid tenant
-  IF ({{SCHEMA_NAME}}.session_tenant IS NULL) THEN
-  	SIGNAL SQLSTATE '99401' SET MESSAGE_TEXT = 'NOT AUTHORIZED: INVALID TENANT ID';
-  END IF;
```
Comment (albertwang-ibm) [inline fhir-persistence-schema/src/main/resources/add_resource_template.sql:65]: I'm wondering, is this store procedure really used?

File: fhir-persistence-schema/src/main/resources/add_resource_type.sql [hunk 4]
```diff
@@ -1,5 +1,5 @@
 -------------------------------------------------------------------------------
--- (C) Copyright IBM Corp. 2016
+-- (C) Copyright IBM Corp. 2016, 2020
```
Comment (albertwang-ibm) [inline fhir-persistence-schema/src/main/resources/add_resource_type.sql:2]: only copyright change?

File: fhir-persistence-schema/src/main/resources/add_any_resource.sql (hunk 1 above)
Comment (prb112) [reply fhir-persistence-schema/src/main/resources/add_any_resource.sql:4]: Yes, as I started with your branch, I think it's an artifact.

File: fhir-persistence-schema/src/main/resources/add_parameter_name.sql (hunk 2 above)
Comment (prb112) [reply fhir-persistence-schema/src/main/resources/add_parameter_name.sql:4]: Yes, as I started with your branch, I think it's an artifact.

File: fhir-persistence-schema/src/main/resources/add_resource_template.sql (hunk 3 above)
Comment (prb112) [reply fhir-persistence-schema/src/main/resources/add_resource_template.sql:65]: I think that was the original and replaced by add_any_resource

2020-04-29 15:05:36.966 00000001    INFO   com.ibm.fhir.schema.app.Main Opening connection to: jdbc:db2://localhost:50000/FHIRDB
2020-04-29 15:05:37.585 00000001    INFO .database.utils.db2.Db2Adapter Create or replace procedure FHIR_ADMIN.SET_TENANT
2020-04-29 15:05:37.648 00000001    INFO .database.utils.db2.Db2Adapter Create or replace procedure FHIRDATA.ADD_CODE_SYSTEM
2020-04-29 15:05:37.722 00000001    INFO .database.utils.db2.Db2Adapter Create or replace procedure FHIRDATA.ADD_PARAMETER_NAME
2020-04-29 15:05:37.752 00000001    INFO .database.utils.db2.Db2Adapter Create or replace procedure FHIRDATA.ADD_RESOURCE_TYPE
2020-04-29 15:05:37.781 00000001    INFO .database.utils.db2.Db2Adapter Create or replace procedure FHIRDATA.ADD_ANY_RESOURCE

File: fhir-persistence-schema/src/main/resources/add_resource_type.sql (hunk 4 above)
Comment (prb112) [reply fhir-persistence-schema/src/main/resources/add_resource_type.sql:2]: right, as I started with the existing branch

**Comment by albertwang-ibm** [comment]
build failed, seems need to merge Lee's PR first

File: fhir-persistence-schema/src/main/resources/set_tenant.sql [hunk 5]
```diff
@@ -20,26 +20,39 @@ BEGIN
...
+      THEN
+        -- If v_not_found, we check only one of the conditions above to see if it is valid
+        -- applications should not be passing invalid keys, so treat this as
+        -- an exception
 
-END
+        -- Check the cause... bad tenant_name
+        SELECT -1 INTO v_not_found
+        FROM fhir_admin.tenants t
+            WHERE t.tenant_name = p_tenant_name;
```
Comment (lmsurpre) [inline fhir-persistence-schema/src/main/resources/set_tenant.sql:41]: sorry, what does SELECT -1 INTO v_not_found do?

Comment (lmsurpre) [reply fhir-persistence-schema/src/main/resources/set_tenant.sql:41]: I guess it only sets if WHERE selects one or more rows?

**Comment by lmsurpre** [comment]
I click re-run on the tests...hopefully works now.
Quick question:  does `set_tenant_unittest.sql` get executed from somewhere or its just for manual testing?

File: fhir-persistence-schema/src/main/resources/set_tenant.sql (hunk 5 above)
Comment (prb112) [reply fhir-persistence-schema/src/main/resources/set_tenant.sql:41]: this is to find out the cause is either tenant_name or other. 
this allows for a differentiated signal back from the stored procedure. 

Yes, that's right

**Comment by prb112** [comment]
> I click re-run on the tests...hopefully works now.
> Quick question: does `set_tenant_unittest.sql` get executed from somewhere or its just for manual testing?

manual tests

**Comment by prb112** [comment]
> I think there is misunderstanding here about the issue, the real issue here is that if the tenant key is set wrongly(**base64 encoded but not the correct one**, if not base64 encoded, then RequestContext codes always throws "Invalid tenant key" exception), then the first create resource request will trigger the correct "Invalid Tenant" exception. but the problem is that the following create resource requests trigger exception like this:
> [4/29/20 23:59:19:741 EDT] 000001ef com.ibm.fhir.persistence.jdbc.dao.impl.ResourceDAOImpl E Failure storing Resource type name id: name=Patient
> com.ibm.db2.jcc.am.SqlIntegrityConstraintViolationException: DB2 SQL Error: SQLCODE=-407, SQLSTATE=23502, SQLERRMC=TBSPACEID=-6, TABLEID=-32766, COLNO=0, DRIVER=4.26.14
> at com.ibm.db2.jcc.am.b7.a(b7.java:806)
> ...
> com.ibm.ws.rsadapter.jdbc.WSJdbcPreparedStatement.execute(WSJdbcPreparedStatement.java:390)
> at com.ibm.fhir.persistence.jdbc.dao.impl.ResourceDAOImpl.readResourceTypeId(ResourceDAOImpl.java:368)
> at com.ibm.fhir.persistence.jdbc.dao.impl.ResourceDAOImpl.insert(ResourceDAOImpl.java:486)
> at com.ibm.fhir.persistence.jdbc.impl.FHIRPersistenceJDBCImpl.create(FHIRPersistenceJDBCImpl.java:264)
> 
> This happens because when processing the following requests, set_tenant is always skipped in get connection because tenant_key variable is null instead of the bad one.

Albert.  In this PR it no longer throws that specific error.  it's more specific. The raised sql exception now has a specific message. I'll throw the exception on the connection and manage that specifically by making the connection null.

File: fhir-persistence-schema/src/main/resources/add_resource_template.sql (hunk 3 above)
Comment (prb112) [reply fhir-persistence-schema/src/main/resources/add_resource_template.sql:65]: I'm going to unresolve, confirm and remove

**Comment by prb112** [comment]
tested with while true; do curl -k -v -X GET -u "fhiruser:*******" -H 'Content-Type: application/fhir+json'   'https://localhost:9443/fhir-server/api/v4/$healthcheck' -H 'X-FHIR-TENANT-ID: default'; sleep 1; done

Confirmed it responded with the appropriate errors

File: fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/AdminSchemaGenerator.java [hunk 6]
```diff
@@ -106,7 +106,10 @@ public void buildSchema(PhysicalDataModel model) {
         this.allTablesComplete.addDependencies(procedureDependencies);
         model.addObject(allTablesComplete);
 
-        model.addProcedure(this.adminSchemaName, SET_TENANT, FhirSchemaConstants.INITIAL_VERSION, () -> SchemaGeneratorUtil.readTemplate(adminSchemaName, adminSchemaName, SET_TENANT.toLowerCase() + ".sql", null), Arrays.asList(allTablesComplete), procedurePrivileges);
+        model.addProcedure(this.adminSchemaName, SET_TENANT, 2,
+                () -> SchemaGeneratorUtil.readTemplate(adminSchemaName, adminSchemaName,
+                        SET_TENANT.toLowerCase() + ".sql", null),
+                Arrays.asList(allTablesComplete), procedurePrivileges);
```
Comment (lmsurpre) [inline fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/AdminSchemaGenerator.java:112]: this worked ok to automatically update the stored procedure as part of `--update-schema`, right?

does it work the same as calling `--update-procs`?

Comment (prb112) [reply fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/AdminSchemaGenerator.java:112]: ack it's in two spots.

Comment (prb112) [reply fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/AdminSchemaGenerator.java:112]: Yes, it's OK. it also happens to be in two spots.

Comment (prb112) [reply fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/AdminSchemaGenerator.java:112]: newest commit adds the second spot

File: fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/FhirSchemaGenerator.java [hunk 7]
```diff
@@ -231,8 +230,7 @@ public void buildAdminSchema(PhysicalDataModel model) {
 
         // The set_tenant procedure can be created after all the admin tables are done
         ProcedureDef setTenant = model.addProcedure(this.adminSchemaName,
-                SET_TENANT,
-                FhirSchemaConstants.INITIAL_VERSION,
+                        SET_TENANT, 2,
```
Comment (lmsurpre) [inline fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/FhirSchemaGenerator.java:233]: 👍

**Pull request merged**